### PR TITLE
FM2-183: Add support for DSTU3

### PR DESCRIPTION
**Pull request by ibacher**
@VaishSiddharth I'm opening this as a draft PR as the communication tools are better integrated than what we've been doing.

Anyone else, we're currently building support for DSTU3 into the module. This will also serve as the pattern for integrating new versions of FHIR into the module, though we may not really be looking to support versions *earlier* that DSTU3.

### CODE REVIEW

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.providers;
+
+import javax.validation.constraints.NotNull;
+
+import java.util.List;
+
+import ca.uhn.fhir.rest.annotation.*;
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java:16]: Minor point, but please configure your IDE not to introduce `*` imports. OpenMRS doesn't like anyone to use them, and it makes some sense.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import ca.uhn.fhir.rest.param.ReferenceAndListParam;
+import ca.uhn.fhir.rest.param.TokenAndListParam;
+import ca.uhn.fhir.rest.param.TokenOrListParam;
+import ca.uhn.fhir.rest.server.IResourceProvider;
+import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException;
+import lombok.AccessLevel;
+import lombok.Setter;
+import org.hl7.fhir.convertors.VersionConvertor_30_40;
+import org.hl7.fhir.dstu3.model.*;
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java:25]: Remove the `*` import

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/ConditionDstu3FhirResourceProvider.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	@SuppressWarnings("unused")
+	public List<Resource> getConditionHistoryById(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Condition condition = conditionService.getConditionByUuid(id.getIdPart());
+		if (condition == null) {
+			throw new ResourceNotFoundException("Could not find condition with Id " + id.getIdPart());
+		}
+		return VersionConvertor_30_40.convertCondition(condition).getContained();
+	}
+	
+	//	// TODO: Required or not?
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/ConditionDstu3FhirResourceProvider.java:69]: Yes! We should implement this!

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportDstu3FhirResourceProvider.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		if (diagnosticReport == null) {
+			throw new ResourceNotFoundException("Could not find diagnosticReport with Id " + id.getIdPart());
+		}
+		
+		return VersionConvertor_30_40.convertDiagnosticReport(diagnosticReport);
+	}
+	
+	// TODO: Required or not?
+	@Create
+	public MethodOutcome createDiagnosticReport(@ResourceParam org.hl7.fhir.r4.model.DiagnosticReport diagnosticReport) {
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportDstu3FhirResourceProvider.java:57]: Required, but we will obviously be given the R3 version of the DiagnosticReport here.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportDstu3FhirResourceProvider.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		return FhirServerUtils.buildCreate(diagnosticReportService.saveDiagnosticReport(diagnosticReport));
+	}
+	
+	// TODO: Required or not?
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportDstu3FhirResourceProvider.java:61]: Required, but we will obviously be given the R3 version of the DiagnosticReport here.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/EncounterDstu3FhirResourceProvider.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	                Patient.SP_FAMILY,
+	                Patient.SP_NAME }, targetTypes = Patient.class) ReferenceAndListParam subjectReference) {
+		return VersionConvertor_30_40.convertBundle(FhirServerUtils.convertSearchResultsToBundle(
+		    encounterService.searchForEncounters(date, location, participantReference, subjectReference)));
+		
+	}
+	
+	@History
+	@SuppressWarnings("unused")
+	public List<Resource> getEncounterHistoryById(@IdParam @NotNull IdType id) {
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/EncounterDstu3FhirResourceProvider.java:74]: I'd prefer if these files were structured similarly, i.e. with the `@History` method following the `@Read` method.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.providers;
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java:10]: Thinking about this some more, I think that instead of naming every `ResourceProvider` `...DStu3ResourceProvider` we should move them into a `r3` package so, for instance, this class would become `org.openmrs.module.fhir2.providers.r3.AllergyIntoleranceFhirResourceProvider`.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import org.hl7.fhir.dstu3.model.*;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.openmrs.module.fhir2.api.FhirAllergyIntoleranceService;
+import org.openmrs.module.fhir2.util.FhirServerUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+@Qualifier("fhirDstu3Resources")
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceDstu3FhirResourceProvider.java:34]: Let's shorten this to `fhirR3Resources` (R3 is the release, but was called STU3 — standard trial use; R2 was released and called DSTU2 — draft standard trial use; I'm just favouring the shorter one!)

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/ObservationDstu3FhirResourceProvider.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	        @OptionalParam(name = Observation.SP_VALUE_STRING) StringAndListParam valueStringParam,
+	        @OptionalParam(name = Observation.SP_DATE) DateRangeParam date,
+	        @OptionalParam(name = Observation.SP_CODE) TokenAndListParam code, @Sort SortSpec sort) {
+		return VersionConvertor_30_40.convertBundle(FhirServerUtils.convertSearchResultsToBundle(
+		    observationService.searchForObservations(encounterReference, patientReference, hasMemberReference, valueConcept,
+		        valueDateParam, valueQuantityParam, valueStringParam, date, code, sort)));
+	}
+	
+	@History
+	public List<Resource> getObservationHistoryById(@IdParam @NotNull IdType id) {
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/ObservationDstu3FhirResourceProvider.java:77]: See above about the order of methods

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/PatientDstu3FhirResourceProvider.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	        @OptionalParam(name = Patient.SP_ADDRESS_POSTALCODE) StringOrListParam postalCode,
+	        @OptionalParam(name = Patient.SP_ADDRESS_COUNTRY) StringOrListParam country, @Sort SortSpec sort) {
+		return VersionConvertor_30_40
+		        .convertBundle(FhirServerUtils.convertSearchResultsToBundle(patientService.searchForPatients(name, given,
+		            family, identifier, gender, birthDate, deathDate, deceased, city, state, postalCode, country, sort)));
+	}
+	
+	@History
+	@SuppressWarnings("unused")
+	public List<Resource> getPatientResourceHistory(@IdParam @NotNull IdType id) {
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/PatientDstu3FhirResourceProvider.java:79]: `getPatientHistoryById`, also ordering.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirDstu3RestServlet.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Collection;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.rest.server.IResourceProvider;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+public class FhirDstu3RestServlet extends FhirRestServlet {
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirDstu3RestServlet.java:21]: To be consistent, let's call this FhirR3RestServlet

File: omod/src/main/java/org/openmrs/module/fhir2/web/filter/ForwardingFilter.java [hunk 12]
```diff
@@ -33,7 +38,29 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
...
-			String newURI = requestURI.replace(prefix, "/ms/fhir2Servlet/");
+			String[] requestPath = requestURI.split(prefix, 1);
+			String maybeVersion = requestPath[1].split("/", 1)[0];
+			
+			String replacement;
+			switch (maybeVersion) {
+				case "R2":
+					prefix += "/" + maybeVersion + "/";
+					replacement = "/ms/fhir2Dstu2Servlet";
+					break;
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/web/filter/ForwardingFilter.java:49]: Let's remove this from this switch since we aren't going to use it right now.

File: omod/src/main/resources/config.xml [hunk 13]
```diff
@@ -50,6 +50,16 @@
...
 
+	<servlet>
+		<servlet-name>fhir2Dstu3Servlet</servlet-name>
+		<servlet-class>org.openmrs.module.fhir2.web.servlet.FhirDstu3RestServlet</servlet-class>
+	</servlet>
+
+	<servlet>
+		<servlet-name>fhir2Dstu2Servlet</servlet-name>
+		<servlet-class>org.openmrs.module.fhir2.web.servlet.FhirDstu2RestServlet</servlet-class>
+	</servlet>
```
Comment (ibacher) [inline omod/src/main/resources/config.xml:61]: We should remove this servlet mapping

File: omod/src/main/resources/config.xml [hunk 14]
```diff
@@ -50,6 +50,16 @@
 		<servlet-class>org.openmrs.module.fhir2.web.servlet.FhirRestServlet</servlet-class>
 	</servlet>
 
+	<servlet>
+		<servlet-name>fhir2Dstu3Servlet</servlet-name>
```
Comment (ibacher) [inline omod/src/main/resources/config.xml:54]: We should change this `servlet-name` to fhir2R3Servlet

File: omod/src/main/java/org/openmrs/module/fhir2/web/filter/ForwardingFilter.java [hunk 15]
```diff
@@ -33,7 +38,29 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
...
+					prefix += "/" + maybeVersion + "/";
+					replacement = "/ms/fhir2Dstu3Servlet";
+					break;
+				case "R4":
+					prefix += "/" + maybeVersion + "/";
+					replacement = "/ms/fhir2Servlet";
+					break;
+				default:
+					replacement = "/ms/fhir2Servlet";
+					break;
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/web/filter/ForwardingFilter.java:60]: We need to rethink this. We should return an error if the value isn't a known version so that we're using a consistent access URL.

File: omod-dstu3/src/test/java/org/openmrs/module/fhir2/web/servlet/BaseFhirResourceProviderTest.java [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+import org.openmrs.module.fhir2.api.impl.FhirGlobalPropertyServiceImpl;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.mock.web.MockHttpServletResponse;
+import org.springframework.mock.web.MockServletConfig;
+import org.springframework.mock.web.MockServletContext;
+import org.springframework.transaction.annotation.Transactional;
+
+public abstract class BaseFhirResourceProviderTest<T extends IResourceProvider, U extends IBaseResource> {
```
Comment (ibacher) [inline omod-dstu3/src/test/java/org/openmrs/module/fhir2/web/servlet/BaseFhirResourceProviderTest.java:51]: Instead of making a fork of this class, maybe it makes more sense to call the version here `BaseFhirR3ResourceProviderTest` and have it extend `BaseFhirResourceProviderTest` and then make the changes that we need here. This might involve changing `BaseFhirResourceProviderTest`, but that's ok.

File: omod-dstu3/src/test/java/org/openmrs/module/fhir2/web/servlet/BaseFhirProvenanceResourceTest.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+import org.hl7.fhir.r4.model.Coding;
+import org.hl7.fhir.r4.model.Practitioner;
+import org.hl7.fhir.r4.model.Provenance;
+import org.hl7.fhir.r4.model.Reference;
+import org.mockito.Mock;
+import org.openmrs.User;
+import org.openmrs.module.fhir2.FhirConstants;
+import org.openmrs.module.fhir2.api.translators.PractitionerReferenceTranslator;
+
+public abstract class BaseFhirProvenanceResourceTest<T extends DomainResource> {
```
Comment (ibacher) [inline omod-dstu3/src/test/java/org/openmrs/module/fhir2/web/servlet/BaseFhirProvenanceResourceTest.java:29]: Instead of making a fork of this class, maybe it makes more sense to call the version here `BaseFhirR3ProvenanceResourceTest` and have it extend `BaseFhirProvenanceResourceTest` and then make the changes that we need here. This might involve changing `BaseFhirProvenanceResourceTest`, but that's ok.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/EncounterDstu3FhirResourceProvider.java (hunk 6 above)
Comment (VaishSiddharth) [reply omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/EncounterDstu3FhirResourceProvider.java:74]: @ibacher This might need to be changed in master as well. I am not changing anything in omod  module because it might be irrelevant for this PR

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/PatientDstu3FhirResourceProvider.java (hunk 10 above)
Comment (VaishSiddharth) [reply omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/PatientDstu3FhirResourceProvider.java:79]: @ibacher This might need to be changed in master as well

Comment (ibacher) [reply omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/PatientDstu3FhirResourceProvider.java:79]: I'm aware. I've got to go through that code soon.

**Comment by ibacher** [comment]
@VaishSiddharth One other thing this would benefit from is rebasing on `master`. We had a change go in to build support for paging, but it requires some pretty sizeable changes.

**Comment by ibacher** [comment]
That's also, incidentally, why the Travis build is failing.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/EncounterFhirResourceProvider.java [hunk 18]
```diff
@@ -53,6 +53,16 @@ public Encounter getEncounterById(@IdParam @NotNull IdType id) {
...
 	
+	@History
+	@SuppressWarnings("unused")
+	public List<Resource> getEncounterHistoryById(@IdParam @NotNull IdType id) {
+		org.hl7.fhir.r4.model.Encounter encounter = encounterService.getEncounterByUuid(id.getIdPart());
+		if (encounter == null) {
+			throw new ResourceNotFoundException("Could not find encounter with Id " + id.getIdPart());
+		}
+		return VersionConvertor_30_40.convertEncounter(encounter).getContained();
+	}
```
Comment (VaishSiddharth) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/EncounterFhirResourceProvider.java:64]: @ibacher Test for all these functions fail. Can you give an idea, what's wrong here

Comment (ibacher) [reply omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/EncounterFhirResourceProvider.java:64]: Likely because the converter doesn't handled the contained resources properly, so when `getContained()` gets called here, it returns an empty list. We may need to look at implementing re-implementing the code to get the provenance resources as part of the R3 submodule.

File: omod/src/test/java/org/openmrs/module/fhir2/providers/FhirResourceProviderTest.java [hunk 19]
```diff
@@ -0,0 +1,10 @@
+package org.openmrs.module.fhir2.providers;
+
+import org.hl7.fhir.instance.model.api.IBaseBundle;
+import org.springframework.mock.web.MockHttpServletResponse;
+
+import java.io.UnsupportedEncodingException;
+
+public interface FhirResourceProviderTest {
+    IBaseBundle readBundleResponse(MockHttpServletResponse response) throws UnsupportedEncodingException;
```
Comment (VaishSiddharth) [inline omod/src/test/java/org/openmrs/module/fhir2/providers/FhirResourceProviderTest.java:9]: @ibacher I have added an interface like this.

Comment (ibacher) [reply omod/src/test/java/org/openmrs/module/fhir2/providers/FhirResourceProviderTest.java]: @VaishSiddharth Sorry... I was apparently leading you astray. [This commit](https://github.com/openmrs/openmrs-module-fhir2/pull/158/commits/88232ddf1a43567bd4b11e0e1cbe88894bbafeee) has the refactoring necessary for this. I left some errors in the DSTU3 test classes though. See if you can work them out.

**Comment by VaishSiddharth** [comment]
@ibacher Not Really :D. I had to manipulate a condition test because 

SP_RECORDED_DATE was not present so changed to SP_ASSERTED_DATE in provider so corresponding test needed some tweaks

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+	public Bundle searchConditions(
+	        @OptionalParam(name = Condition.SP_PATIENT, chainWhitelist = { "", Patient.SP_IDENTIFIER, Patient.SP_NAME,
+	                Patient.SP_GIVEN, Patient.SP_FAMILY }) ReferenceAndListParam patientParam,
+	        @OptionalParam(name = Condition.SP_SUBJECT, chainWhitelist = { "", Patient.SP_IDENTIFIER, Patient.SP_NAME,
+	                Patient.SP_GIVEN, Patient.SP_FAMILY }) ReferenceAndListParam subjectParam,
+	        @OptionalParam(name = Condition.SP_CODE) TokenAndListParam code,
+	        @OptionalParam(name = Condition.SP_CLINICAL_STATUS) TokenAndListParam clinicalStatus,
+	        @OptionalParam(name = Condition.SP_ONSET_DATE) DateRangeParam onsetDate,
+	        @OptionalParam(name = Condition.SP_ONSET_AGE) QuantityAndListParam onsetAge,
+	        //SP_RECORDED_DATE not present so changed to SP_ASSERTED_DATE
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ConditionFhirResourceProvider.java:88]: This is correct.

File: omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProvider.java [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+		return VersionConvertor_30_40.convertObservation(observation).getContained();
+	}
+	
+	@Search
+	public IBundleProvider searchObservations(
+	        @OptionalParam(name = Observation.SP_ENCOUNTER) ReferenceAndListParam encounterReference,
+	        @OptionalParam(name = Observation.SP_SUBJECT, chainWhitelist = { "", Patient.SP_IDENTIFIER, Patient.SP_GIVEN,
+	                Patient.SP_FAMILY,
+	                Patient.SP_NAME }, targetTypes = Patient.class) ReferenceAndListParam patientReference,
+	        //TODO: SP_HAS_MEMBER not present so converted to SP_BASED_ON just to prevent errors
```
Comment (ibacher) [inline omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProvider.java:71]: This should be `SP_RELATED`, but note that `SP_RELATED` has a somewhat different structure.

Comment (VaishSiddharth) [reply omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProvider.java:71]: @ibacher Shouldn't this be `SP_RELATED_TYPE` [here](http://hl7.org/fhir/STU3/observation.html#search) is why I think so.
Also only changing the data type will not solve the problems, I think that we'll have to map it, as we did for `Task` also the `observation` resource returns `IBundleProvider` in search function which complicates things or if not all the tests are failing with this error message
```
ca.uhn.fhir.context.ConfigurationException: This context is for FHIR version "DSTU3" but the class "org.hl7.fhir.r4.model.OperationOutcome" is for version "R4"

```
I guess we'll need to override this method also in `BaseFhirR3ResourceProviderTest`

Comment (ibacher) [reply omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProvider.java:71]: No. `SP_RELATED`. `SP_RELATED_TYPE` refers to the type structure that is part of the related element. Technically, `SP_RELATED_TARGET` is the more direct replacement for `SP_HAS_MEMBER`, but only when the `related` structure looks like this:

```json
"related": {
    "type": "has-member",
    "target": "Observation/3"
}
```

Or, in terms of search parameters:

```
/Observation?related.type=has-member&related.target=Observation/3
```

This also isn't exactly like Task, because what we're converting here isn't a full resource, but a search parameter.

The message you highlight:

```
ca.uhn.fhir.context.ConfigurationException: This context is for FHIR version "DSTU3" but the class "org.hl7.fhir.r4.model.OperationOutcome" is for version "R4"
```

Does indicate a problem, but part of the problem is that we're not properly handling things when the result is an `OperationOutcome` (which itself needs to be translated). Please step through one of these failing tests with a debugger to see what is actually happening.

> I guess we'll need to override this method also in BaseFhirR3ResourceProviderTest

Which method are you referring to?

Comment (VaishSiddharth) [reply omod-dstu3/src/main/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProvider.java:71]: > Which method are you referring to?


protected void describeMismatchSafely(MockHttpServletResponse item, Description mismatchDescription) in this function we have `OperationOutcome` which is of `import org.hl7.fhir.r4.model.OperationOutcome;` type so I think this needs to be generic so that all versions and use it

**Comment by VaishSiddharth** [comment]
@ibacher Thanks for your commit. I was trying to do something similar to solve the error :+1:

**Comment by VaishSiddharth** [comment]
@ibacher I have tested the latest commit with Postman and both endpoints work fine. :+1:

File: omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirVersionUtils.java [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+		R4,
+		R3,
+		ERROR
+	}
+	
+	public static Enum<FhirVersion> getFhirResourceVersion(HttpServletRequest request) {
+		String requestURI = request.getRequestURI();
+		String contextPath = request.getContextPath();
+		String prefix = contextPath + "/ws/fhir2";
+		if (requestURI.startsWith(prefix)) {
```
Comment (VaishSiddharth) [inline omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirVersionUtils.java:26]: @ibacher I get a NPE here when I do something like this

```java
return request.getScheme() + "://" + request.getServerName() + port + "/" + contextPath + "/ws/fhir2/"
		        + FhirVersionUtils.getFhirResourceVersion(request);
```
in `OpenmrsFhirAddressStrategy`

File: omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirVersionUtils.java [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+package org.openmrs.module.fhir2.web.servlet;
+
+import javax.servlet.http.HttpServletRequest;
+
+public class FhirVersionUtils {
+	
+	public enum FhirVersion {
+		R4,
+		R3,
+		ERROR
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirVersionUtils.java:19]: I would make this `UNKNOWN`. Also I would order the values in the enum this way: `UNKNOWN`, `R3`, `R4`. That way we're adding values to the end and (hopefully) not messing up the ordinal order of the enum.

File: omod/src/main/java/org/openmrs/module/fhir2/web/filter/ForwardingFilter.java [hunk 24]
```diff
@@ -39,45 +40,26 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
...
-						
-						String newURI = requestURI.replace(prefix, replacement);
-						req.getRequestDispatcher(newURI).forward(req, res);
-						return;
-					}
-				}
+			String fhirVersion = String.valueOf(FhirVersionUtils.getFhirResourceVersion(request));
+			
+			String replacement;
+			switch (fhirVersion) {
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/web/filter/ForwardingFilter.java:46]: It would be better to switch on the enum value here, even if we do need to calculate the string value. That way we have a compile-time check of the various cases.

File: omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirVersionUtils.java (hunk 22 above)
Comment (ibacher) [reply omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirVersionUtils.java:26]: Can you give me some more context for this? What does the new version of `OpenmrsFhirAddressStrategy` look like? What does the test case that you are using look like?

**Comment by ibacher** [comment]
@VaishSiddharth I've rebased this on the latest changes and squashed things down.
